lib/traject/macros/stanford.rb

```diff
@@ -0,0 +1,96 @@
+# frozen_string_literal: true
+
+module Macros
+  # Macros for extracting Stanford Specific MODS values from Nokogiri documents
+  module Stanford
+    NS = { mods: 'http://www.loc.gov/mods/v3',
+           rdf: 'http://www.w3.org/1999/02/22-rdf-syntax-ns#',
+           dc: 'http://purl.org/dc/elements/1.1/' }.freeze
+    # SOURCE SPECIFIC METHODS
+    # Macros for extracting Stanford-specific values from Nokogiri documents
+    def generate_sul_manifest
+      lambda { |record, accumulator, context|
+        druid = generate_druid(record, context)
+        # eventually add a test that calls IIIF service to check
+        if druid.present? && druid?(druid.chomp('druid:'))
+          manifest = "https://purl.stanford.edu/#{druid}/iiif/manifest"
+          accumulator << manifest
+          grab_sul_iiif_links(manifest)
+          generate_sul_shown_at(druid)
+        end
+      }
+    end
+
+    def generate_druid(record, context)
+      if context.settings.fetch('identifier')
+        identifier = context.settings.fetch('identifier').dup
+        identifier.sub! 'stanford_', ''
+      else
+        record.xpath('/*/mods:identifier', NS).map(&:text)
+      end
+    end
+
+    def druid?(identifier)
+      identifier =~ /([a-z]{2})(\d{3})([a-z]{2})(\d{4})\z/
+    end
+
+    def generate_sul_shown_at(druid)
+      to_field 'agg_is_shown_at' do |record, accumulator|
+        mods_url = record.xpath('/*/mods:location/mods:url', NS).map(&:text)
+        accumulator.concat(mods_url) if mods_url.present?
+        accumulator << "https://purl.stanford.edu/#{druid}"
+      end
+    end
+
+    def grab_sul_iiif_links(manifest)
+      result = Net::HTTP.get(URI.parse(manifest))
```

Comment:
Instead of using `Net::HTTP` directly, I think we should explicitly add `faraday` as a dependency and use that. It's a little more flexible and handles some basic work for you... I think it'll also let us check whether the response was a success instead of throwing exceptions.